### ISSUE TITLE
page query not working for posts page

### ISSUE THREAD

**Issue by mephysto** [wp-graphql/wp-graphql#2486]
### Description

As of 1.9.x I can not longer grab certain attibutes of the posts page using the page query

### Steps to reproduce

1. Use WPGraphQL 1.9.x
2. Set up a page (i.e., "news") to be the posts page in the Reading settings (/wp-admin/options-reading.php)
3. Edit this page. Add a featured image via the sidebar.
4. Attempt to retrieve this image using following query: 
```
{
  page(idType: URI, id: "/news/") {
    title
    isPostsPage
    featuredImage {
      node {
        alt: altText
        src: sourceUrl
      }
    }
  }
}
```
5. In GraphQL explorer, I would get the following error:
```
 "errors": [
    {
      "debugMessage": "Argument 1 passed to WPGraphQL\\Type\\InterfaceType\\NodeWithFeaturedImage::WPGraphQL\\Type\\InterfaceType\\{closure}() must be an instance of WPGraphQL\\Model\\Post, instance of WPGraphQL\\Model\\PostType given, called in /nas/content/live/example/wp-content/plugins/wp-graphql/src/Type/WPConnectionType.php on line 447",
      "message": "Internal server error",
      "extensions": {
        "category": "internal"
      },
      "locations": [
        {
          "line": 5,
          "column": 5
        }
      ],
      "path": [
        "page",
        "featuredImage"
      ]
    }
  ],
  "data": {
    "page": {
      "title": null,
      "isPostsPage": false,
      "featuredImage": null
    }
  },
  ```
6. If I downgrade to WPGraphQL version 1.8.x, I get my intended results:
```
  "data": {
    "page": {
      "title": "My Cool News Stories",
      "isPostsPage": true,
      "featuredImage": {
        "node": {
          "alt": "aerial-view-of-boat",
          "src": "https://www.example.com/wp-content/uploads/2020/09/aerial-view-of-boat.jpg"
        }
      }
    }
  },
  ```
7. 

### WPGraphQL Version

1.9.x

### WordPress Version

6.0.1

### PHP Version

7.4

### Additional enviornment details

Attempted to:
- disable all other plugins except WPGraphQL, problem persisted.
- I checked if Image used still existed in Media Library, and not a weird mismatch or missing image

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have disabled ALL plugins except for WPGraphQL.

- [X] Yes
- [ ] My issue is with compatibility with a specific WordPress plugin, and I have listed all my installed plugins (and version info) above.

**Comment by justlevine** [comment]
[👀 slack thread](https://wp-graphql.slack.com/archives/C3NM1M291/p1660788916272649?thread_ts=1660788916.272649&cid=C3NM1M291)

**Comment by jasonbahl** [comment]
@mephysto ok, I've confirmed this bug. 

![CleanShot 2022-08-30 at 11 00 47](https://user-images.githubusercontent.com/1260765/187496526-687f8ed4-d76f-41b8-9b19-51d4592aa38a.png)

The problem here is that the page shouldn't be returning at all when making this query, instead, it should be returning null. 

If we create a page and name it "Blog"

![CleanShot 2022-08-30 at 11 05 58](https://user-images.githubusercontent.com/1260765/187497503-30a0e51e-1951-4dc6-a79c-d72e941d5951.png)

Then set that page to be the "page for posts" 

![CleanShot 2022-08-30 at 11 16 14](https://user-images.githubusercontent.com/1260765/187500860-26d883d7-2c2f-4213-b80f-e01431bda84f.png)

Then, visit the url in the WordPress-rendered front-end. 

We will see that WordPress does not return a page: 

![CleanShot 2022-08-30 at 11 19 20](https://user-images.githubusercontent.com/1260765/187501928-aef4a996-6b2a-4ac3-9408-901cb349a35e.png)

In the screenshot above, I show the body class, which would have `page-*` classes if it were considered a page.

We can also use the following snippet to test this: 

```php
add_action( 'template_redirect', function() {
	wp_send_json( [
		'isPage' => is_page()
	]);
} );
```

And visiting the `/blog` url again will show that this url does not represent a page:
  
![CleanShot 2022-08-30 at 11 22 18](https://user-images.githubusercontent.com/1260765/187503050-2355a5a7-abd0-475c-a8fc-ccb0020ddd09.png)

If we change the settings back to where there is no "Posts Page" set:

![CleanShot 2022-08-30 at 11 20 10](https://user-images.githubusercontent.com/1260765/187502337-c161b5f8-0552-4cbc-a7b3-f938da00d640.png)

And then visit the same URL, we will now see that WordPress returns a Page for this path. 

![CleanShot 2022-08-30 at 11 20 47](https://user-images.githubusercontent.com/1260765/187502553-5e7bd574-bf56-41ec-9a17-e49a516d7178.png)

And with the debug snippet we can see the url does indeed represent a Page at this point. 

![CleanShot 2022-08-30 at 11 22 02](https://user-images.githubusercontent.com/1260765/187502938-d42d3f60-e5a2-4858-9ad3-1dde04b2b03e.png)

-----

All that said, the expected behavior _should_ be for the `page( uri: "/news" ) { ... }` query to return null when the News page is set as the "Posts Page"

If we expand the query further and try to query for `page(id: "/blog", idType: URI) ' and `nodeByUri(uri:"/blog")` at the same time, we will see the problem clearly: 

![type-mismatch](https://user-images.githubusercontent.com/1260765/187504388-8580c300-4a94-420a-99b6-626a8e08933f.png)

In the screenshot here, we see that the `page(id: "/blog", idType: URI)` query has returned a node with `__typename: 'Page'` and the `nodeByUri(uri:"/blog")` query has returned a `__typename: "ContentType"`. 

If we enter the graph from 2 different places, and try to resolve the same URI, we should get the same thing. 

WordPress has already dictated that this URI, when set as the Posts Page, should not be a page, so querying a page _should_ return null. 

The expected behavior, to match core WordPress behavior, _should_ be as follows:

![CleanShot 2022-08-30 at 11 32 11](https://user-images.githubusercontent.com/1260765/187504843-7645eea2-bfd1-4b9b-b8ee-1dcf5b5d6a71.png)

Querying for a page directly will return null, because that uri doesn't represent a page.

Querying for `nodeByUri` will return a `ContentType` because that url represents the archive of posts for the blog.

**Comment by jasonbahl** [comment]
I've opened a PR here to fix this bug: https://github.com/wp-graphql/wp-graphql/pull/2490

Now, this of course does cause a limitation. 

If the URI is now returning a `ContentType` instead of a `Page`, how can users query meta / custom fields associated with the page that is indeed the "Page for Posts"?

I think for now, the best thing to do would be to register a field to the `ContentType` type that returns the data you need, if the `ContentType` is the "Post" ContentType. 

Something like the following: 

```php
add_action( 'graphql_register_types', function() {

	register_graphql_field( 'ContentType', 'pageForPosts', [
		'type'       => 'Page',
		'descripton' => __( 'Query the Page For Posts from the ContentType', 'your-text-domain' ),
		'resolve'    => function( $content_type ) {
			if ( $content_type->name !== get_post_type_object( 'post' )->name ) {
				return null;
			}
			$page_id = get_option( 'page_for_posts', 0 );

			if ( empty( $page_id ) ) {
				return null;
			}

			$page = get_post( (int) $page_id );

			return $page instanceof WP_Post ? new \WPGraphQL\Model\Post( $page ) : null;
		}
	] );

});
```

That would allow you to query something like so, where you can get fields from the Page from the `ContentType` that's associated with the Page.

![CleanShot 2022-08-30 at 12 15 56](https://user-images.githubusercontent.com/1260765/187512771-0e10ea48-b736-458c-9e04-73c72612b739.png)

And the field will return null for Content Types that are not the "Post" post type, for example querying for the archive of a "movies" custom post type:

![CleanShot 2022-08-30 at 12 15 47](https://user-images.githubusercontent.com/1260765/187512954-2c74fa6b-92f2-4290-863f-c1fe9fb8c037.png)

**Comment by justlevine** [comment]
> WordPress has already dictated that this URI, when set as the Posts Page, should not be a page, so querying a page _should_ return null.

The problem is that WP doesn't care as much as GraphQL about polluting types, and -edit: as you noted in your follow up  - treats the Posts Page as both a page and a content type.

This is core WP functionality, and I don't think it's the right move to remove it and force users to code around it.  Especially considering that this used to work.

Perhaps we could work around this by resolving some sort of `ContentArchive` intermediary type / interface to provide an alternative way to access the `Page`-related data on the `ContentType`?

**Comment by jasonbahl** [comment]
@justlevine ya, even in WordPress core though, you have to do some workarounds to get mixed behavior between the archive for posts and the posts page. Because, as noted, WordPress doesn't recognize this path as a page anymore, so globals are different, etc. I know folks have a lot of solutions for doing things with a mix of the "Page" and the "archive" but they're all creative solutions already and not "Core" behavior.

I definitely thing the right move is to work toward interfaces for Archives, as suggested.

I think a better Interface approach here could actually lead to better support for this scenario than core WordPress has itself, where we can better surface (and document in the Schema) when Archives and Pages are related, etc.

I think we should work on that in a separate PR. 

This addresses a bug with the current behavior. I know some users have been using this bug as a feature, but the intent has always been to have the behavior I outlined, and this corrects that bug. 

The snippet I provided should help users in the meantime as we work toward a better solution for using Interfaces to expose Archives for content types.

**Comment by jasonbahl** [comment]
> is it that much different than querying post or page by the same slug resulting in two different types?

@justlevine we only allow querying by slug on non-hierarchical post types because of this. 

Hierarchical post types can have multiple objects in the same post type with the same slug. 

For example, I could have pages like: 

- company
  - about
- products
  - about

And while the paths (uri) would be `/company/about` and `/products/about`, both slugs would be `about`. 

This is why WPGraphQL doesn't allow for `{ page( id: "about", idType: SLUG ) { ... }}` and there's also not a `nodeBySlug` entry point that lets you query across post types using the slug as a unique identifier, because it's not a unique identifier.

**Comment by jasonbahl** [comment]
@justlevine I opened an issue to track the Interface approach for this. 

https://github.com/wp-graphql/wp-graphql/issues/2491#issue-1356229434

**Comment by justlevine** [comment]
> I know folks have a lot of solutions for doing things with a mix of the "Page" and the "archive" but they're all creative solutions already and not "Core" behavior.

This particular GH issue ( featured images, but it's not the only property) documents core behavior, however, and the object returned by WP isn't synonymous with a WP_Taxonomy either. 

It's also a very common usage pattern (iirc employed by some starters), and forcing what's essentially a breaking change on users without a core-supported alternative makes me extremely uneasy. Especially when 
1. the problem this solution addresses is a schema hygiene issue, and not a pressing issue to users.
2. The proposed workaround doesn't revert the breaking-ish change, but adds a field which will eventually require another frontend-breaking rewrite when an actual solution is integrated into core.

I strongly recommend we address this _for now_ by reverting to pre-1.9.x behavior, and merge #2490 only after WPGraphQL has a solution for exposing the page-related fields that are missing from `ContentType`.

**Comment by jasonbahl** [comment]
> This particular GH issue ( featured images, but it's not the only property) documents core behavior, however, and the object returned by WP isn't synonymous with a WP_Taxonomy either.

Users can still query the page that is set as the posts page, they just can't identify it by the URI, because that URI identifies something else. 

With my fix, a user could still create a page, set it as "Posts Page" then write a query for that page using the ID. 

For example, I can still query the page, and its featured image if I can identify it by a unique Identifier that applies to it: 

![CleanShot 2022-08-30 at 14 49 51](https://user-images.githubusercontent.com/1260765/187540495-a2178ac6-f6ea-49d1-9fab-bdc327e2ad96.png)

But, I appropriately should _not_ be able to query it by URI, because when a page is set as the "Posts Page" the URI doesn't represent the page, but the archive. 

The intent of a URI is to be a Unique Resource Identifier (an ID), and if WPGraphQL is resolving different things with the same URI, that's a bug and this PR aims to fix that. 

I think we should continue fixing bugs, but for folks that are using the bug as a feature, we can provide snippets to let them carry on until they have time to update to alternative approaches. 

Here's a filter that brings back this functionality:

```php
add_filter( 'graphql_pre_resolve_uri', function( $null, $uri, $context, $wp ) {

	$page_for_posts = get_option( 'page_for_posts', 0 );

	if ( empty( $page_for_posts ) ) {
		return $null;
	}

	$permalink = get_permalink( (int) $page_for_posts );

	if ( empty( $permalink ) || is_wp_error( $permalink ) ) {
		return $null;
	}

	$parsed_permalink = wp_parse_url( $permalink );
	$parsed_uri = wp_parse_url( $uri );

	if ( ! isset( $parsed_permalink['path'] ) || ! isset( $parsed_uri['path'] )) {
		return $null;
	}

	$trimmed_uri_path = rtrim( ltrim( $parsed_uri['path'], '/' ), '/' );
	$trimmed_permalink_path = rtrim( ltrim( $parsed_permalink['path'], '/' ), '/' );

	if ( $trimmed_permalink_path === $trimmed_uri_path ) {
		return new \WPGraphQL\Model\Post( get_post( (int) $page_for_posts ) );
	}

	return $null;

}, 10, 4 );
```

**BEFORE THIS FILTER (bug fixed)**

![CleanShot 2022-08-30 at 14 48 57](https://user-images.githubusercontent.com/1260765/187543250-f9f51272-20f0-4580-90c4-9ade2046789f.png)


**AFTER THIS FILTER (old behavior restored)**

![CleanShot 2022-08-30 at 15 07 06](https://user-images.githubusercontent.com/1260765/187543343-40a9d6f6-8bb5-4f6e-9ea1-1f875e25df18.png)

**Comment by jasonbahl** [comment]
I updated the [v1.9.0 Release Notes](https://github.com/wp-graphql/wp-graphql/releases/tag/v1.9.0) to include this snippet as well, and a link to this discussion.

**Comment by richardcalahan** [comment]
Thanks for this snippet. Curious what an alternative approach would look like without the snippet? 

I'm confused by WP not returning a "page" when it's assigned to page_for_posts. It can still hav all the same data (content, featured image, SEO if using yoast) so it's still a page in basic terms. 

Without this snippet, the ContentType returned can't get any of that info. What's a recommended approach?

**Comment by justlevine** [comment]
@richardcalahan , as mentioned above, you can still get the page_for_posts `page` by querying by its `id`, just not by its `uri`. 

In the future, either `ContentType`s will likely have either an `Archive` with the page data or queries by `uri` will resolve to an intermediate `Archive` type with both the page and ContentNode date.

We'll be exploring this in #2491, feel free to weigh in over there 😎

**Comment by havredevers** [comment]
Thanks for all your job.
It should be nice if we could get this page_for_posts  with a targeted query. For now, I have to query all pages and then filter the result.

This could be something like : 
```
query GET_POSTS_PAGE {
  pages(where: {isPostsPage: true}) {
    nodes {
      isPostsPage
      id
      title
      slug
    }
  }
}
```

**Comment by keyfilter** [comment]
Here’s a valuable insight https://phps.space/questions/what-could-be-causing-my-getoption-wordpress-function-to-return-a-null-value-each-time-i-refresh-the-page
